test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +462,288 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
   EXPECT_EQ(output.getTailSlots()[0], mOp2);
 }
 
+// Tests that a min sized instruction held at the end of the fetch buffer is
+// allowed to be predecoded in the same cycle as being fetched
+TEST_P(PipelineFetchUnitTest, minSizeInstructionAtEndOfBuffer) {
+  ON_CALL(isa, getMaxInstructionSize()).WillByDefault(Return(insnMaxSizeBytes));
+  ON_CALL(isa, getMinInstructionSize()).WillByDefault(Return(insnMinSizeBytes));
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+  // Buffer will contain valid min size instruction so predecode returns
+  // min bytes read
+  MacroOp mOp = {uopPtr};
+  ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(insnMinSizeBytes)));
+
+  // Fetch the data, only min bytes will be copied to fetch buffer. Should allow
+  // continuation to predecode
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  uint64_t setPC = blockSize - insnMinSizeBytes;
+  // Fetch a single minimum sized instruction, buffered bytes = 0
+  fetchUnit.updatePC(setPC);
+  // Tick. Fetch data and predecode
+  fetchUnit.tick();
+
+  // Buffer should now be empty as all bytes predecoded
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 0);
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp);
+
+  // Expect a block starting at address 16 to be requested when we fetch again
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory,
+              requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+      .Times(1);
+  fetchUnit.requestFromPC();
+
+  // Tick again, expecting that decoding will now resume
+  MacroOp mOp2 = {uopPtr2};
+  memory::MemoryReadResult nextBlockValue = {{16, blockSize}, 0, 1};
+  span<memory::MemoryReadResult> nextBlock = {&nextBlockValue, 1};
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(nextBlock));
+  ON_CALL(isa, predecode(_, _, _, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp2), Return(insnMaxSizeBytes)));
+
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  // Completed reads called again as more data is requested
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  // Output of width 1 so only 1 call to predecode
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  fetchUnit.tick();
+
+  // Initially 0 bytes, 16 bytes added, max bytes predecoded leaving (16 - max)
+  // bytes left
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 16 - insnMaxSizeBytes);
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp2);
```

Comment:
I think rather than address 16, we should base the next fetch address off of the block size

Reply:
This comes under the TODO for blockSize parameterisation and follows the previous implementation

Reply:
Applied to other comments